### DISCUSSION TITLE
ggcirclepack, now with counting - but not 100%

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#15] [Show and tell]
Have returned to the circle packing, now with an option for counting and it mostly works, except when faceting. 
 https://github.com/EvaMaeRey/ggcirclepack

I do conditional grouping to propogate aes channels that would otherwise drop out in counting, which seems a little awkward but kind of works.  I'll probably take another look at this soon.  

```r
compute_panel_circlepack <- function(data, scales, count = F){

  if(is.null(data$area)){data$area <- 1}
  
    if(is.null(data$area)){data$area <- 1}
  
   # if(!("weight" %in% names(data))){data$weight <- 1}
  # order matters... Need to add text aesthetics
  if("id" %in% names(data)){data <- group_by(data, id, .add = T)}

  if("fill" %in% names(data)){data <- group_by(data, fill, .add = T)}
  if("alpha" %in% names(data)){data <- group_by(data, alpha, .add = T)}
  if("colour" %in% names(data)){data <- group_by(data, colour, .add = T)}
  if("group" %in% names(data)){data <- group_by(data, group, .add = T)}
  if("linetype" %in% names(data)){data <- group_by(data, linetype, .add = T)}
  if("linewidth" %in% names(data)){data <- group_by(data, linewidth, .add = T)}
  
  if(count){
  data %>% 
    mutate(id = as.integer(as.factor(id))) %>% 
    count(wt = area) %>% 
    rename(area = n) ->
    data
  } else { data$id = 1:nrow(data)}
  
  data %>%
    pull(area) %>%
    packcircles::circleProgressiveLayout(
      sizetype = 'area') %>%
    packcircles::circleLayoutVertices(npoints = 50) %>%
    left_join(data) # should add join by id

}


```

With count turned on, you can have area represent count, (here the number of countries in each continent)

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/5be60c5f-8daf-44a7-a105-1abea84eaee0)

Overwriting id to country, we get unpacking to country.  Text is messy but it's working. 

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/a1855d47-dc95-4574-a9cb-5541ac264045)

Setting the area to represent pop, we get something nicer (default aes for size is after_stat(area))

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/369fcf28-66d6-4b1c-a5c4-94a2348dcf23)

But facetting, reveals a breakdown in how things are supposed to work...

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/8ccd4e46-02b1-49a3-b554-a2d73d8be57f)

If you have count = F, in the layers, though, you can still get a nice result:

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/414d8ee5-747c-4df8-a4a8-3505a32f0d34)

**Comment by yjunechoe** [comment]
I know nothing about circlepacking and grouping aesthetics, but a dplyr tangent:

I think you can simplify the conditional and incremental `group_by()`s with `pick(any_of(...))`, like:

```r
grp_cols <- c("id", "fill", "alpha", ...)

data <- data %>%
  group_by(pick(any_of(grp_cols)))
```

And this should also preserve the order of columns specified in the external vector

**Reply by EvaMaeRey** [reply]
I ended up doing this...

```r
data_mapped_aes_names <- names(data)[names(data) %in% c("id", "fill", "alpha", 
                                             "colour", "group", "linewidth", 
                                             "linetype")]
    
  data %>% 
    group_by(across(data_mapped_aes_names)) ->
  data 
```

... but your pick-any of looks great and I will use that instead!  Thank you!


I like this debugging version - circles become flags (npoints = 3)!  
![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/3ae961b2-d792-46c1-9b01-0a7edf14e0e9)

**Reply by EvaMaeRey** [reply]
![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/7efd8098-7b7b-470f-ba4a-d91737e8d847)
And facet is working!

**Comment by EvaMaeRey** [comment]
To wrap up. This seemed to be the offending piece:   mutate(id = as.integer(as.factor(id)))  , and we end up computing id as row number for the count version (to join onto computed xy positions for circle).

The rationale to add counting is allows us to fluidly disaggregate (or just do more computation *within* your plot pipeline) as you can do with other layers like statcount/ geombar!

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/a578d18b-e748-41c9-adb0-15a022abc2b7)
![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/cb154714-0e6a-4e2b-b236-755b322299ba)
![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/161f8bae-3204-4921-aaef-a10dcd3de2ce)
![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/cff90fd3-0bd1-4531-a462-03339dccd525)
![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/5fa11d34-14db-4f9a-9573-3f7a8113f571)

Seems to be working, but obviously early stages.

**Reply by EvaMaeRey** [reply]
Also cool to do would be to generalize and do summary on a variable (value) instead of only count. Default would be count - fun is sum and value is 1 if null.

**Comment by EvaMaeRey** [comment]
Nothing new to report, just enjoying ggcirclepack w/ LIsa's Garden data - what an amazing garden.  Doing first ever + aes(fill = I("blue")) kind of declaration and really like it!

``` r
library(tidyverse)

harvest_2021 <- readr::read_csv('https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2024/2024-05-28/harvest_2021.csv')

spending_2021 <- readr::read_csv('https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2024/2024-05-28/spending_2021.csv')


spending_2021 %>% 
  ggplot() + 
  labs(title = "Lisa's Garden Spending, 2021") +
  aes(area = price, id = "All") + 
  ggcirclepack::geom_circlepack(show.legend = F, 
                                color = "black", 
                                linewidth = .15) + 
  coord_equal() + 
  ggcirclepack::geom_circlepack_text(show.legend = F) +
  ggstamp::theme_void_fill()  + 
  aes(label = after_stat(paste0(id, "\n$", area))) + 
  aes(fill = I("aliceblue"))
```

![](https://i.imgur.com/AwAL3NB.png)

``` r

last_plot() +
  aes(id = vegetable) + 
  labs(title = "Spending breakdown") + 
  aes(label = after_stat(paste0(id, "\n$", area)))
```

![](https://i.imgur.com/JUYpM5B.png)

``` r

last_plot() %+% 
  harvest_2021 +
  aes(area = weight) + 
  labs(title = "Lisa's Garden 2021, Harvest") + 
  aes(label = after_stat(paste0(id, "\n", round(area/1000), "kg"))) +
  aes(fill = I("lightsalmon3")) + 
  aes(color = I("grey90"))
```

![](https://i.imgur.com/d6ZENej.png)

``` r

last_plot() +
  aes(id = "All") + 
  labs(title = "Harvest, Aggregate") 
```

![](https://i.imgur.com/xt9aJZS.png)

<sup>Created on 2024-06-06 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by EvaMaeRey** [comment]
Finally reaching out to author @mbedward, I learned that that project was motivate by actual tree packing dynamics. https://github.com/mbedward/packcircles uses a pairwise repulsion packing algorithm.   'It had the very nice property of distributing small circles around and in between larger circles. This mirrors the typical arrangement of young and mature trees in the Australian woodlands'  The algorithm is also used to model soap bubble arrangements.

I was also referred to https://github.com/mbedward/packcircles/issues/3, in which a complementary geom was suggested.  :-)  

Updating the initial geom, (aggregations must be performed by the user), mostly inspired by the geom_bar() capabilities - you could move as flexibly between plots with geom_bar, but arguably bars are just not as charming, especially after the nth bar chart.  
https://evamaerey.github.io/mytidytuesday/2024-06-06-lisas-garden/flipbook/garden_flipbook.html#33

**Reply by EvaMaeRey** [reply]
Changing out circlepacking geoms with geom bar, composition has strong parallels.  
![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/84bd9a75-f915-42ad-8e0b-2510fe2fa92d)

**Comment by EvaMaeRey** [comment]
Circle packing in the wild:  https://github.com/gkaramanis/tidytuesday/blob/master/2024/2024-week_24/pride_index.R  :-)    Some feedback here might be that label layer is a challenge to use directly and to get good looking results.

**Reply by EvaMaeRey** [reply]
Based on the conditional circle packing labeling strategy, and thinking about ways to make this easier for users...

1.   adding a `render` aes - based on the condition, should the circle be labeled or not.  
2.  And making a stat version of geom_circlepack_text so you can use GeomTextRepel.   
  
 I'd be interested in any reactions.  the new aes(render = var_TF_condition) is not really demoed in the working package, and maybe there are reasons to reverse course.    

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/1b07f5e7-d9b5-4aae-af92-c851732a98a4)


![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/8dacf1d8-3892-4d57-b900-fa3df2ff9448)

**Reply by teunbrand** [reply]
A few random thoughts:

* Isn't the `render` aesthetic just an alias for `label = ifelse(test, label_var, NA)`?
* Does the text repelling take into account the size of the circle?

**Reply by EvaMaeRey** [reply]
Good questions.   🤔🤔

> Does the text repelling take into account the size of the circle?

geom_circlepack_text  labels at the center of the circle.  StatCirclepackcenter is the stat and it returns 1 row per circle with x y as center and a radius.    stat_circlepack_center() with ggrepel, doesn't look at the radius.   I believe it repel just does it's thing as it would on a scatterplot of circle centers (not knowing the packing context). 

> Isn't the render aesthetic just an alias for label = ifelse(test, label_var, NA)

This is worth thinking about.  I have a default aes is used aes(label = after_stat(id)).  I would like that default to be usable and then for user to decied 'too much labeling', let's focus.  If using the ifelse NA strategywith repulsion, will the NA's would be in play in terms of placement? Yes right? Maybe that'd indeed yield nicer placement results. 

Thanks!

**Reply by teunbrand** [reply]
Yeah perhaps `label = after_stat(ifelse(test, id, NA))` becomes a little bit too verbose and `render` is a good shorthand.

>  will the NA's would be in play in terms of placement

I think for `geom_text_repel()` yes as they count as a point to be dodged (but the text itself needn't be).

**Reply by EvaMaeRey** [reply]
I tried something like this as the last step in compute panel, for unlabeled circle center dodging in the case of ggrepel usage, but I'm not sure it's better (I think it's a little harder to manage) and my go back to row deletion if render is false.  

```
 if (!is.null(data$render)) {
        data <- data %>% mutate(auto_label = ifelse(.data$render, 
             as.character(.data$id), NA))
     }
     else {
         data <- data %>% mutate(auto_label = id)
     }
     data

```

**Reply by EvaMaeRey** [reply]
Yet another approach may be to use size...
  
`scale_size(range = c(-3, 5))`

does not throw an error.  Below zero size doesn't seem to render (treated like zero).  But close to zero and positive renders and is distracting since you can't really read it.  It would be nice to have a cliff transformation.  e.g. map to consistent decrease from 5 to 2 and then suddenly 0 in size for smaller sizes.  It's not something I've seen though.

**Reply by yjunechoe** [reply]
You could hack this with `stage()`! I see that the default aes for your Stat ggproto is `size = after_stat(area)`. So the user can do something like `size = stage(after_stat = area, after_scale = ifelse(size < 2, 0, size)` and then add `... + scale_size(range = c(0,5))` to their plot - this should give them the "cliff" behavior for free.

For a more dynamic/user-friendly approach, I think the Stat should take an argument for specifying the lower-bound for `size` to force values to zero, and store that in `params`. I think you can abuse `Stat$finish_layer()` for applying this transformation, since `size` is available at that step.

I was also considering whether you could write an actual "transformer" for this with `trans_new()` (for use in `scale_*(transform = )`, but I'm don't think scale transformations know drawing (vs. data) units to express something like "set small text sizes to zero"

**Reply by EvaMaeRey** [reply]
Cliff behavior for free sounds pretty great!   Maybe it could be wrapped up into some nice-to-use shortcut in the package, and try it out for a while.

The dynamic/user-friendly approach is something to think about.  But I feel like even default_aes setting is a little mysterious and oversteps a 'just leave the defaults as they are' philosophy which seems more maintainer-friendly.  (I also just not confident of syntax patterns get all that done - maybe this is the truer heart of my reluctance). but... https://mitchelloharawild.com/blog/ggplot2-package-design/ 'I refuse to add extra options for changing the style of the plot even if it is easy to do so.'

**Reply by EvaMaeRey** [reply]
Re transformer, I'm liking stage idea best, so also will just keep it in mind!  Thanks!

**Reply by EvaMaeRey** [reply]
Here is a ggnewscale approach to consistent relative size differences for label types,

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/4bcbfc24-c3c1-4189-b943-8d1659acb590)

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/bb04c703-c376-4981-b063-f5b388870a4d)

 but Elio reminded me that delayed aes evaluation could work well in this case (since it is just a multiplier).  I think it's another case that requires stage. 

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/45da4bad-d879-4ea7-a942-5ee3efb5c87a)

**Reply by EvaMaeRey** [reply]
I realize that I didn't actually allow size to be directly mapped (because it wasn't one of the grouped vars, so dropped out upon summararize():

```
data_mapped_aes_names <- names(data)[names(data) %in% c("id", "fill", "alpha", 
                                             "colour", "group", "linewidth", 
                                             "linetype")]
    
  data %>% 
    group_by(across(data_mapped_aes_names)) ->
  data 
```

I've since added label and size as grouping vars... but maybe I need a more systematic approach, like I should be grouping on all mapped variables except for the area variable and the weight variable...

**Comment by EvaMaeRey** [comment]
Generalizing, so aggregation can not only be count, but also average (or something else). 

First demos...

Average highway mpg ...

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/29cd5ab9-40d5-4a2c-9c3c-af9b28381aa1)

Average pop over 50 years ...

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/a705f7f5-5619-4d8c-88fb-2eb09aa4a7b7)

**Reply by EvaMaeRey** [reply]
I have all these 'Unknown or uninitialised column' warnings I think because of how I'm adding columns if the column doesn't exist 

e.g.   data$wt <- 1  

my inclination is to try a mutate function to add the new variable, but it seems like it would be relatively verbose.

**Reply by EvaMaeRey** [reply]
There's a 'stat chaining' feel to this.  I think ideally it would be good to break out and stabilize the aggregation part of the compute and make it generally available for other area charts like waterfall, wedge, georegion etc...  W regard to stabilizing, I've been trying to capture all the possible grouping variables to make this work, but I think looking at the complement (what not to group by) might be the way to go.  Which would be area and weight, I think in this case.   Some discussion in #20 got me thinking about this possibility.

**Reply by teunbrand** [reply]
> 'Unknown or uninitialised column'

These are tibble warnings that don't happen with base data.frames.

> There's a 'stat chaining' feel to this

There has been a lot of discussion on this lately and I wonder what a `StatChain` would look like.

> I've been trying to capture all the possible grouping variables

Is there a reason to deviate from ggplot2's automatic grouping?

**Reply by EvaMaeRey** [reply]
> These are tibble warnings that don't happen with base data.frames.

Would I have to switch out my dplyr manipulation?  

> There has been a lot of discussion on this lately and I wonder what a `StatChain` would look like.

I'm not sure though I think https://mjskay.github.io/ggbuilder/ probably worth a look.

I think for this project step in the right direction is to just break up the compute functions, and then bring them together in a single function to define the Stat's compute_*. 

> Is there a reason to deviate from ggplot2's automatic grouping?

I specified my Stat w compute_panel, to work alongside GeomPolygon, and then auto-grouping doesn't happen. But maybe rewriting for use w/ GeomCircle and defining compute_group, we'd be in a better place to use the automatic grouping.  A Circlepack Stat without aggregation, designed to work with GeomPolygon is #54 - open to suggestions!

**Comment by EvaMaeRey** [comment]
Another @gkaramanis example! Hooray! https://github.com/gkaramanis/tidytuesday/tree/master/2024/2024-week_28

I like the comparison of the two area charts - I think this makes a case for circle packing being part of the exploratory data visualization tool kit.    w/ bar charts (probably the default for count per category), when there are a lot of categories (as is the case for the number of function david robinson uses), you kind of need to make the filtering move, but circle packing lets you leave all the categories and circle size draws your attention to the big players.   You don't *have to* subset w/ circlepacking (though you certainly can).  @gkaramanis also commented to me that @wilkox's ggfittext might be worth a look.   https://github.com/wilkox/ggfittext

**Reply by wilkox** [reply]
Great idea for a package! ggfittext can't handle circles, but it seemed like an interesting problem so I got a [basic solution working](https://github.com/wilkox/ggcircumscribe). I'll keep working on it and let you know if/when it might be worth importing.

**Reply by EvaMaeRey** [reply]
This looks incredibly promising - glad that you're interested and have the know-how to put something together so quickly!  treemappify definitely played a role in wanting to work on a geom function for circle packing, by the way.

**Reply by EvaMaeRey** [reply]
Working with the text sizing from ggcircumscribe.  Comparin ggcirclepack::geom_circlepack_text() with geom_circumscribe() in a case where radii are all the same.     geom_circumscribe grow result is very nice!  


``` r
library(tidyverse)

harvest_2021 <- readr::read_csv('https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2024/2024-05-28/harvest_2021.csv')
#> Rows: 726 Columns: 5
#> ── Column specification ────────────────────────────────────────────────────────
#> Delimiter: ","
#> chr  (3): vegetable, variety, units
#> dbl  (1): weight
#> date (1): date
#> 
#> ℹ Use `spec()` to retrieve the full column specification for this data.
#> ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
```

``` r

harvest_2021 %>% 
  distinct(vegetable) %>% 
  ggplot() + 
  aes(id = vegetable) + 
  ggcirclepack::geom_circlepack() + 
  ggcirclepack::geom_circlepack_text() + 
  coord_equal() + 
  aes(fill = I("aliceblue")) + 
  aes(color = I("midnightblue"))

```

![](https://i.imgur.com/d3BM5ug.png)

``` r
  

layer_data(i = 2) ->
circle_data

```

``` r

library(ggcircumscribe)

ggwipe::last_plot_wipe_last() +   
  ggcircumscribe::geom_circumscribe(data = circle_data, 
                    aes(id = NULL, x= x, y = y, label = label), 
                    radius = grid::unit(10, "mm"), grow = TRUE)

#> ! The `reflow` argument is not yet implemented
```

![](https://i.imgur.com/DTNAAp3.png)

<sup>Created on 2024-08-03 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by EvaMaeRey** [comment]
Area charts: circlepack v ggplot2::geom_bar. 

![image](https://github.com/user-attachments/assets/ef1fd22b-48b8-482d-94e6-ffd1e2cb422b)

from https://www.linkedin.com/posts/srivatsan-srinivasan-b8131b_justforfun-activity-7225513781362094080-oswF?utm_source=share&utm_medium=member_desktop

**Comment by EvaMaeRey** [comment]
An initial attempt at adding aggregation compute is now active in 'sf2stat'!  This is a little harder space because I think aggregations (count, weighted count, average) should be mapped to fill, and not a non-native aes area as in circlepack.  For the most part, I think this can work, only a little discouraged by adjusting scales with fill:

The problem is seen in this example where outerspace_objects has a record of number launched objects for each country and each year.  So if we want to show total number of objects for each country, we can 1) aggregate before plotting, or 2) use the aggregation computation. But, applying new fill scale will be unhelpful in approach 2, which I think is because the scale is registered early on in the build (and we declare fill directly and then compute to a new fill value).  😔  

Aggregating before plotting: 

![image](https://github.com/user-attachments/assets/0abecb2c-93bd-46f7-9ff6-5258c63bfc98)

Aggregating within plot:

![image](https://github.com/user-attachments/assets/cdc35b09-2d93-457f-9a57-5030c010d3ef)


Maybe if the user enters num_objects as the wt instead of fill, we will get the intended result, but it feels a little burdensome to think through.

**Reply by teunbrand** [reply]
I've not thought this through at all, but the 2nd colourbar leads me to think the log10 transform is applied incorrectly. Perhaps it is transformed twice or inverted from data space instead of transformed space?

**Reply by EvaMaeRey** [reply]
I will add some more details when I get a chance to have another look.  There are some other puzzles the new aggregation to figure out as well - but they seem like they might be easier/possible fixes.

**Reply by EvaMaeRey** [reply]
Minimal example of this behavior (fill -> aggregated fill)

 Say we want to look at the populations of north and south carolina and north and south dakota, where we have household records recorded.  We can see 'depth' of the population by jittering.  Or we can aggregate (jittering is not really viable for maps).  But it could be nice if the layer's Stat takes care of that.   However w/ log scale, fill that is aggregated seems to break.

``` r 
library(tidyverse)

set.seed(12346)

tibble(x = sample(c("North", "South"), 100, replace = T, prob = c(.3,.7)), 
       y = sample(c("Carolina", "Dakota"), 100, replace = T, prob = c(.3,.7)),
       household_size = sample(1:7, 100, replace = T, prob = 7:1)) ->
df

df
## # A tibble: 100 × 3
##    x     y        household_size
##    <chr> <chr>             <int>
##  1 North Dakota                3
##  2 South Dakota                2
##  3 South Dakota                5
##  4 South Carolina              3
##  5 North Carolina              4
##  6 North Dakota                2
##  7 South Carolina              6
##  8 South Dakota                1
##  9 South Dakota                7
## 10 South Carolina              2
## # ℹ 90 more rows

df |> 
  summarise(agg = sum(household_size), .by = c(x,y))

## # A tibble: 4 × 3
##   x     y          agg
##   <chr> <chr>    <int>
## 1 North Dakota      70
## 2 South Dakota     169
## 3 South Carolina    72
## 4 North Carolina    23
```

## jittering

``` r
ggplot(df) + 
  aes(x = x, y = y, fill = household_size) + 
  geom_point(shape = 21, size = 15, position = "jitter") + 
  labs(title = "'depth' is shown via jitter")
```

![image](https://github.com/user-attachments/assets/fb5fe2bb-d068-4737-992f-327f32bf0caf)

## aggregate 

``` r
compute_fill_collapse <- function(data, scales){
  
  if(is.null(data$fill)){data$fill <- 1}
  
  data |> 
    summarise(fill = sum(fill),
              summary = sum(fill))
  
}

StatFillCollapse <- ggproto("StatFillCollapse", 
                            Stat,
                            compute_group = compute_fill_collapse,
                            default_aes = aes(fill = after_stat(stage(fill))))

# deep fill, collapsing by group
ggplot(df) + 
  aes(x = x, y = y, fill = household_size) + 
  geom_point(stat = StatFillCollapse, shape = 21, size = 15)
```

![image](https://github.com/user-attachments/assets/a98d920c-0e4e-4d44-854b-be5acaa415ad)

``` r
layer_data()
##      fill summary x y PANEL group shape colour size alpha stroke
## 1 #132B43      23 1 1     1     1    21  black   15    NA    0.5
## 2 #275378      70 1 2     1     2    21  black   15    NA    0.5
## 3 #28547B      72 2 1     1     3    21  black   15    NA    0.5
## 4 #56B1F7     169 2 2     1     4    21  black   15    NA    0.5
```


# applying log transformation

``` r
# uh-oh, log transformation not doing what might want
last_plot() + 
  scale_fill_viridis_c(transform = "log10")

```

![image](https://github.com/user-attachments/assets/dc2471e3-27a8-484d-81a4-5790bd91edbe)

``` r
layer_data()
##      fill   summary x y PANEL group shape colour size alpha stroke
## 1 #440154  3.107210 1 1     1     1    21  black   15    NA    0.5
## 2 #37678C  9.417096 1 2     1     2    21  black   15    NA    0.5
## 3 #36688C  9.512071 2 1     1     3    21  black   15    NA    0.5
## 4 #FDE725 22.017303 2 2     1     4    21  black   15    NA    0.5
```

**Reply by EvaMaeRey** [reply]
Not sure if I should have the expectation that this *could* work...
